File: src/main/java/net/csibio/aird/compressor/sortedintcomp/IntegratedVarByteWrapper.java
/*
 * Copyright (c) 2020 CSi Biotech
 * Aird and AirdPro are licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 */

package net.csibio.aird.compressor.sortedintcomp;

import me.lemire.integercompression.differential.IntegratedIntCompressor;
import me.lemire.integercompression.differential.IntegratedVariableByte;
import net.csibio.aird.enums.SortedIntCompType;

public class IntegratedVarByteWrapper implements SortedIntComp {

    @Override
    public String getName() {
        return SortedIntCompType.IVB.getName();
    }

    /**
     * Delta+VByte+ByteCompressor Encoder. Default ByteCompressor is Zlib. The compress target must be
     * sorted integers.
     *
     * @param uncompressed the sorted integers
     * @return the compressed data
     */
    @Override
    public int[] encode(int[] uncompressed) {
        int[] compressedInts = new IntegratedIntCompressor(new IntegratedVariableByte()).compress(uncompressed);
        return compressedInts;
    }

    @Override
    public int[] decode(int[] compressed) {
        int[] uncompressed = new IntegratedIntCompressor(new IntegratedVariableByte()).uncompress(compressed);
        return uncompressed;
    }
}

File: CSharpSDK/Compressor/SortedIntComp/IntegratedVarByteWrapper.cs
/*
 * Copyright (c) 2020 CSi Studio
 * AirdSDK and AirdPro are licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions of the Mulan PSL v2. 
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2 
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.  
 * See the Mulan PSL v2 for more details.
 */

using CSharpFastPFOR.Differential;

namespace AirdSDK.Compressor
{
    public class IntegratedVarByteWrapper : SortedIntComp
    {
        //使用VariableByte算法将排序了的int数组进行压缩
        public override string getName()
        {
            return SortedIntCompType.IVB.ToString();
        }

        public override int[] encode(int[] uncompressed)
        {
            int[] compressed = new IntegratedIntCompressor(new IntegratedVariableByte()).compress(uncompressed);
            return compressed;
        }

        //使用VariableByte算法对已经压缩的int数组进行解压缩
        public override int[] decode(int[] compressed)
        {
            int[] uncompressed = new IntegratedIntCompressor(new IntegratedVariableByte()).uncompress(compressed);
            return uncompressed;
        }
    }
}